mmengine/fileio/io.py

```diff
@@ -15,7 +52,735 @@
+        's3'
+
+    Returns:
+        str: Return the prefix of uri if the uri contains '://'. Otherwise,
+        return ''.
+    """
+    assert is_filepath(uri)
+    uri = str(uri)
+    # if uri does not contains '://', the uri will be handled by
+    # HardDiskBackend by default
+    if '://' not in uri:
+        return ''
+    else:
+        prefix, _ = uri.split('://')
+        # In the case of PetrelBackend, the prefix may contain the cluster
+        # name like clusterName:s3://path/of/your/file
+        if ':' in prefix:
+            _, prefix = prefix.split(':')
+        return prefix
+
+
+def _get_file_backend(prefix: str, backend_args: dict):
+    """Return a file backend based on the prefix or backend_args.
+
+    Args:
+        prefix (str): Prefix of uri.
+        backend_args (dict): Arguments to instantiate the corresponding
+            backend.
+    """
+    # backend name has a higher priority
+    if 'backend' in backend_args:
+        backend_name = backend_args.pop('backend')
+        backend = backends[backend_name](**backend_args)
+    else:
+        backend = prefix_to_backends[prefix](**backend_args)
+    return backend
+
+
+def get_file_backend(
+    uri: Union[str, Path, None] = None,
+    *,
+    backend_args: Optional[dict] = None,
+    enable_singleton: bool = False,
+):
+    """Return a file backend based on the prefix of uri or backend_args.
+
+    Args:
+        uri (str or Path): Uri to be parsed that contains the file prefix.
+        backend_args (dict, optional): Arguments to instantiate the
+            corresponding backend. Defaults to None.
+        enable_singleton (bool): Whether to enable the singleton pattern.
+            If it is True, the backend created will be reused if the
+            signature is same with the previous one. Defaults to False.
+
+    Returns:
+        BaseStorageBackend: Instantiated Backend object.
+
+    Examples:
+        >>> # get file backend based on the prefix of uri
+        >>> uri = 's3://path/of/your/file'
+        >>> backend = get_file_backend(uri)
+        >>> # get file backend based on the backend_args
+        >>> backend = get_file_backend(backend_args={'backend': 'petrel'})
+        >>> # backend name has a higher priority if 'backend' in backend_args
+        >>> backend = get_file_backend(uri, backend_args={'backend': 'petrel'})
+    """
+    global backend_instances
+
+    if backend_args is None:
+        backend_args = {}
+
+    if uri is None and 'backend' not in backend_args:
+        raise ValueError(
+            'uri should not be None when "backend" does not exist in '
+            'backend_args')
+
+    if uri is not None:
+        prefix = _parse_uri_prefix(uri)
+    else:
+        prefix = ''
+
+    if enable_singleton:
+        # TODO: whether to pass sort_key to json.dumps
+        unique_key = f'{prefix}:{json.dumps(backend_args)}'
+        if unique_key in backend_instances:
+            return backend_instances[unique_key]
```

Comment:
I think it's good to pass `sort_key` to `json.dumps`

Reply:
If `backend_args` contains string key and integer key, passing `sort_keys` to `json.dumps` will throw a `TypeError`.

```python
>>> import json
>>> json.dumps(a, sort_keys=True)
'[11, 222]'
>>> a = {'a1': [11], 1: [13]}
>>> json.dumps(a, sort_keys=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/developer/miniconda3/envs/py37-pt120-cu116/lib/python3.7/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/home/developer/miniconda3/envs/py37-pt120-cu116/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/developer/miniconda3/envs/py37-pt120-cu116/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
TypeError: '<' not supported between instances of 'int' and 'str'
>>> json.dumps(a)
'{"a1": [11], "1": [13]}'
```

Reply:
Current implementation will create multiple instances if arguments are in different orders.
e.g. `args1 = {'k1': 1, 'k2': 2}` and `args2 = {'k2': 2, 'k1': 1}`
Maybe we should work out a method to sort `type` first and then `elements` with the same type. Is builtin function `id` feasible?

Reply:
The reason to use singleton is to avoid creating a lot of instances repeatedly but it is acceptable to create multiple instances. Therefore, it may not be necessary to handle it this way.